service/matching/taskReader.go

```diff
@@ -123,22 +125,21 @@ dispatchLoop:
 			}
 			task := newInternalTask(taskInfo, tr.tlMgr.completeTask, enumsspb.TASK_SOURCE_DB_BACKLOG, "", false)
 			for ctx.Err() == nil {
-				// We checked if the task was expired before putting it in the buffer, but it
-				// might have expired while it sat in the buffer, so we should check again.
-				if taskqueue.IsTaskExpired(taskInfo) {
+				if valid := tr.taskValidator.maybeValidate(taskInfo, tr.tlMgr.taskQueueID.taskType); !valid {
 					task.finish(nil)
 					tr.taggedMetricsHandler().Counter(metrics.ExpiredTasksPerTaskQueueCounter.GetMetricName()).Record(1)
 					// Don't try to set read level here because it may have been advanced already.
-					break
+					continue dispatchLoop
```

Comment:
I really think we should do some preparatory refactoring here. What this code is doing doesn't actually seem complicated, but because we have nested loops, we're doing all these unconventional things like continuing to a label, not doing if err != nil early returns, and cancel statements in weird places. I think this makes it harder to find bugs. For example, the `for ctx.Err() == nil` condition is redundant because we're selecting on `case <-ctx.Done()` anyway. Who knows what else could be wrong here. It's hard enough to parse without even looking for issues.

Reply:
totally agree that (probably) the entire matching service should be refactored. 
but this is out the scope of this PR.

if you plan to refactor the matching, consider use [this](https://github.com/temporalio/temporal/pull/4141) as reference